pkg/cache/cache.go

```diff
@@ -19,17 +19,20 @@ import (
 )
 
 var (
-	ErrNotFound = errors.New("not found")
+	ErrNotFound      = errors.New("not found")
+	ErrUnimplemented = errors.New("unimplemented")
 )
 
 // Getter wraps a method to read from cache.
 type Getter interface {
 	Get(key interface{}) (interface{}, error)
+	GetAll() ([]interface{}, error)
 }
 
 // Putter wraps a method to write to cache.
 type Putter interface {
 	Put(key interface{}, value interface{}) error
+	PutHash(key interface{}, value interface{}) error
```

Comment:
Not sure why do we need this? I think we can reuse `Put` function.

Reply:
Just in case we want to use `RedisHashCache.Put(...)` interface, I think we should keep the existing `Put` implementation for RedisHashCache. 🤔

Reply:
> Just in case we want to use RedisHashCache.Put(...) interface

Can you explain more about this? That is for putting a set of (field, value) for a key?

Reply:
> That is for putting a set of (field, value) for a key? 

Since the current `RedisHashCache` has embedded `RedisCache` instance, in the caller side, we can use `RedisHashCache.Get(key)` which call to `RedisHashCache.RedisCache.Get(key)` with the current implementation. In that case, It will return an error not found if we override the `Put` interface with our `PutHash` implementation`. That why I think we should keep the existing `Put` interface and add `PutHash` which be used for the new purpose of storing (k,v) as values of the defined hashkey of RedisHashCache.

Reply:
I think we should not embed  `RedisCache`. We can either add a new `cache: RedisCache` field or use `redis.Redis` directly.
I think we have no reason to put `PutHash` function to all cache interfaces.

Reply:
>I think we have no reason to put PutHash function to all cache interfaces.

I feel the same way with @nghialv

Reply:
Just implementing `Put()` to `RedisHashCache` isn't enough?

Reply:
Well, if I understand correctly, I feel like the current RedisCache is enough. We can just use piped-key as a key and use a stats byte sequence as a value. Tell me why using HSET if any.

Reply:
I think that is why `GetAll` and `HRedis` were introduced. We can have all stats of all pipeds in a single command in a fast way to expose for prometheus.

Reply:
🙆‍♀️  Get you guys' point 🙏 
Addressed on https://github.com/pipe-cd/pipe/pull/2190/commits/478ee0d1c9d6dfa9114511cf278d81206ffc4977 👀 
At this implementation, I change all implementation for Cache interface of RedisHashCache as it will treat the Hash object stored in redis as its backend and replace all `XXX` commands with `HXXX` commands.

Reply:
alright, makes sense. it's for retrieving all stats in bulk.

Reply:
> Tell me why using HSET if any.

With only the single `SET` interface, we need a way to know all possible piped (ID) currently registered to the controlplane and query for the data one by one which much slower, I guess 👀